### PR TITLE
NN-3006 damage obligations api

### PR DESCRIPTION
**Pull request by JamieDraperUK**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+@Builder
+@Getter
+@ApiModel(description = "Damage obligation for an offender")
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class OffenderDamageObligationModel {
+
+    @ApiModelProperty(value = "Identifier of damage obligation", example = "1", position = 1)
+    private Long id;
+
+    @ApiModelProperty(value = "Offender number", example = "A12345", position = 2)
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java:22]: what is this field?  Would be good to put an example in the correct format e.g. G5033UT

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    private String offenderNo;
+
+    @ApiModelProperty(value = "Reference number", example = "012388", position = 3)
+    private String referenceNumber;
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java:26]: Is this a `reference` or a reference number?  The example is all digits but leading zero - good to know the correct format?

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    @ApiModelProperty(value = "Amount paid", example = "10.0", position = 8)
+    private BigDecimal amountPaid;
+
+    @ApiModelProperty(value = "Status", example = "ACTIVE", position = 9)
+    private String status;
+
+    @ApiModelProperty(value = "Comment", example = "Damages to canteen furniture", position = 10)
+    private String comment;
+
+    @ApiModelProperty(required = true, value = "Currency of these amounts.", example = "GBP", position = 11)
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java:49]: is this the only field that is `required`?

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java [hunk 4]
```diff
@@ -201,4 +204,15 @@ CaseNote updateOffenderCaseNote(@ApiParam(value = "Noms ID or Prisoner number (a
...
             @ApiParam(value = "Toggle to return IEP detail entries in response (or not).", required = true) @RequestParam(value = "withDetails", required = false, defaultValue = "false") boolean withDetails);
+
+
+    @GetMapping("/{offenderNo}/damage-obligations")
+    @ApiOperation(value = "Return a list of damage obligations")
+    @ApiResponses(value = {
+            @ApiResponse(code = 404, message = "Offender does not exists or is in a different caseload to the user", response = ErrorResponse.class),
+            @ApiResponse(code = 500, message = "Unrecoverable error occurred whilst processing request.", response = ErrorResponse.class)})
+    OffenderDamageObligationResponse getOffenderDamageObligations(
+            @ApiParam(value = "offenderNo", required = true, example = "A1234AA") @PathVariable("offenderNo") @NotNull String offenderNo,
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java:215]: missing `final`?

Comment (petergphillips) [reply src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java:215]: oops ignore me - didn't realise it was an interface

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderResourceImpl.java [hunk 5]
```diff
@@ -231,4 +234,13 @@ public PrivilegeSummary getLatestBookingIEPSummaryForOffender(final String offen
         Optional.ofNullable(booking).orElseThrow(EntityNotFoundException.withId(offenderNo));
         return bookingService.getBookingIEPSummary(booking.getBookingId(), withDetails);
     }
+
+    @Override
+    public OffenderDamageObligationResponse getOffenderDamageObligations(final String offenderNo, final String status) {
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderResourceImpl.java:239]: no `VerifyOffenderAccess`?

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationResponse.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+import lombok.Builder;
+import lombok.Data;
+
+import java.util.List;
+
+@Builder
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationResponse.java:11]: would be better with a constructor, can't see the point of a builder with one argument

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderResourceImpl.java (hunk 5 above)
Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderResourceImpl.java:239]: I've placed it on the service. Would you prefer it on the Resource instead?

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    @ApiModelProperty(value = "Prison the damages occurred", example = "MDI", position = 6)
+    private String prisonId;
+
+    @ApiModelProperty(value = "Amount left to pay", example = "50.0", position = 7)
+    private BigDecimal amountToPay;
+
+    @ApiModelProperty(value = "Amount paid", example = "10.0", position = 8)
+    private BigDecimal amountPaid;
+
+    @ApiModelProperty(value = "Status", example = "ACTIVE", position = 9)
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java:43]: Is this a reference code?  Would be good to list all possible statuses and maybe even reference code field that it comes from?

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java (hunk 3 above)
Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java:49]: Copy and paste error, will fix now.

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderDamageObligationService.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+public class OffenderDamageObligationService {
+    private final OffenderDamageObligationRepository repository;
+    private final String currency;
+
+    public OffenderDamageObligationService(@Value("${api.currency:GBP}") final String currency, final OffenderDamageObligationRepository repository) {
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/service/OffenderDamageObligationService.java:19]: not sure why currency is a constructor parameter - can't see it ever changing.  Better to just use field initialisation

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderDamageObligationService.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        this.repository = repository;
+        this.currency = currency;
+    }
+
+    @VerifyOffenderAccess
+    public List<OffenderDamageObligationModel> getDamageObligations(final String offenderNo, final String status) {
+
+        try {
+            final var damages = StringUtils.isNotEmpty(status) ?
+                    repository.findOffenderDamageObligationByOffender_NomsIdAndStatus(offenderNo, status).stream() :
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/service/OffenderDamageObligationService.java:29]: not sure why `.stream()` is here - it is on both anyway, would be better to move to the `return` statement

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java (hunk 2 above)
Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java:26]: This is indeed a reference number. I'll do some digging to figure out if there is a special format.

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderResourceImpl.java (hunk 5 above)
Comment (petergphillips) [reply src/main/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderResourceImpl.java:239]: just spotted it lower down :grinning:

**Review by petergphillips — APPROVED** [review]
LGTM

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java (hunk 7 above)
Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java:43]: I'll try get access to preprod to gather all statuses.

File: .circleci/config.yml [hunk 10]
```diff
@@ -12,8 +12,7 @@ executors:
       - image: circleci/openjdk:11-jdk-browsers
     working_directory: ~/app
     environment:
-      JAVA_TOOL_OPTIONS: -Xmx1024m -XX:ConcGCThreads=2 -XX:ParallelGCThreads=2 -Djava.util.concurrent.ForkJoinPool.common.parallelism=2
-
+      _JAVA_OPTIONS: -Xmx1024m -XX:ParallelGCThreads=2 -XX:ConcGCThreads=2 -XX:ParallelGCThreads=2 -Djava.util.concurrent.ForkJoinPool.common.parallelism=2 -Dorg.gradle.daemon=false -Dorg.gradle.jvmargs=-XX:+UseContainerSupport
```
Comment (JamieDraperUK) [inline .circleci/config.yml:15]: I tried 512m previously with no luck so upped it to 1024m which seems to work.

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class OffenderDamageObligationModel {
+
+    @ApiModelProperty(value = "Identifier of damage obligation", example = "1", position = 1)
+    private Long id;
+
+    @ApiModelProperty(value = "Offender number", example = "G4346UT", position = 2)
+    private String offenderNo;
+
+    @ApiModelProperty(value = "Reference number", example = "841177/1, A841821/1, 842371", position = 3)
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderDamageObligationModel.java:25]: I'm not fully sure about the format at the moment. So for now I've added some examples of real world ones.

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java [hunk 12]
```diff
@@ -201,4 +204,15 @@ CaseNote updateOffenderCaseNote(@ApiParam(value = "Noms ID or Prisoner number (a
...
+            @ApiParam(value = "Filter by obligation status. Leave blank to return all", required = false, example = "ACTIVE", allowableValues = "INACT,PAID,ONH,ACTIVE,APPEAL") @RequestParam(value = "status", required = false, defaultValue = "") String status);
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java:216]: This is complete list of obligation status values pulled from preprod.

**Review by petergphillips — APPROVED** [review]
LGTM

**Pull request merged**